usb: device_next: Add MIDI 1.0 support to the USB next MIDI 2.0 driver

nrbrook opened the pull request:
usb: device_next: Add MIDI 1.0 support to the USB next MIDI 2.0 driver

Added MIDI 1.0 support to the USB next MIDI 2.0 driver.

Currently, this implements one input and one output jack. I did start implementing the configuration of these through device tree, but got lost down a macro rabbit hole, so figured I should get some feedback first.

The interfaces are selected using `int usbd_midi_set_mode(const struct device *dev, bool enable_midi1, bool enable_midi2)`, to allow for midi 1 or midi 2 to be optionally disabled (default is both enabled).

It also adds a `usbd_midi_send_midi1` function and `rx_midi1_cb` callback for sending and receiving raw midi 1 data. These use the currently active altsetting, converting data formats if needed.

@titouanc

nrbrook commented on include/zephyr/usb/class/usbd_midi2.h:
```diff
@@ -36,6 +37,19 @@ struct usbd_midi_ops {
+	 * MIDI 2.0 alternate is active, @ref rx_packet_cb is used instead.
+	 */
+	void (*rx_midi1_cb)(const struct device *dev, uint8_t cable_num, const uint8_t *bytes,
+			    uint8_t len);
```
Yes, it does already (determined by the parameters to init). It builds the usb descriptor based on the Boolean values

tmon-nordic replied:
I wasn't thinking about duplicating, but rather splitting it in a way that it can be used separately or combined.

nrbrook replied:
I think the easiest way to do that would be via Kconfig, keeping everything in the same file with conditional compilation, unless you have another suggestion?

titouanc commented:
As a general remark, could you please provide a sample application, or update the existing [`samples/subsys/usb/midi`](https://github.com/zephyrproject-rtos/zephyr/tree/v4.3.0/samples/subsys/usb/midi) to show how you intend to use the USB-MIDI1 interface, and possibly how it relates to the USB-MIDI2 ?

In particular:
- Is the USB-MIDI1 interface supposed to work as a fallback to the MIDI2 interface, such that hosts without MIDI2 support can use the device ?
- How is the USB-MIDI1 topology (input/output jacks) configured ?
- Does the application switch between USB-MIDI1/2 ?

Furthermore, I would advise to send/receive data in [Universal MIDI Packets](https://docs.zephyrproject.org/latest/doxygen/html/group__midi__ump.html) over the USB-MIDI1 interface, to keep the API consistent and interoperable with other MIDI transports (network, USB-MIDI2). For example:

```c
int usbd_midi1_send(const struct device *dev, const struct midi_ump ump)
{
	if (UMP_MT(ump) == UMP_MT_MIDI2_CHANNEL_VOICE) {
		// MIDI2 messages not supported on MIDI1 interface
		return -EINVAL;
	}
	// ...
}
```

nrbrook commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -23,34 +26,61 @@ LOG_MODULE_REGISTER(usbd_midi2, CONFIG_USBD_MIDI2_LOG_LEVEL);
 
 /* midi20 A.1 MS Class-Specific Interface Descriptor Types */
-#define CS_GR_TRM_BLOCK	0x26
+#define CS_GR_TRM_BLOCK 0x26
```
Again, clang format

ehughes commented:
@nrbrook @titouanc 

I tested with an NXP i.MX RT685 target and verify it works w/ Win11.

tmon-nordic commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -153,9 +219,314 @@ struct usbd_midi_data {
+		return -EINVAL;
+	}
+
+	if (status < MIDI_STATUS_SYSEX_START) {
```
I don't quite understand what is the reason behind these if statements. Is it some sort of optimization taking advantage of particular numerical value properties?

tmon-nordic commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -227,26 +242,47 @@ struct usbd_midi_data {
+	if (alt == MIDI2_ALTERNATE) {
+		return true;
+	}
+#endif
```
In general you should avoid `#if` inside function body if the compiler can compile-time optimize things away. This is the case here. Just use `if (IS_ENABLED(CONFIG_USBD_MIDI2_ALTSETTING_MIDI1) && alt == MIDI1_ALTERNATE)` and compiler will remove the code away.

The difference is that code under `#if` can easily be missed during refactor, leading to non-compilable Kconfig combinations.

tmon-nordic reviewed (commented):
Please reduce the number of `#if IS_ENABLED()` in favor of compile time `if (IS_ENABLED() &&`

nrbrook commented:
I would like to have the label removed

jfischer-no commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -423,18 +860,21 @@ static void usbd_midi_rx_work(struct k_work *work)
+			LOG_DBG("Rx enqueue requested while device is disabled");
+		} else {
+			LOG_ERR("Failed to enqueue Rx net_buf -> %d", ret);
+		}
```
These changes do not seem to belong in this commit.

jfischer-no commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -400,8 +838,7 @@ static uint8_t usbd_midi_get_bulk_out(struct usbd_class_data *const class_data)
 
-	if (USBD_SUPPORTS_HIGH_SPEED &&
-	    usbd_bus_speed(uds_ctx) == USBD_SPEED_HS) {
+	if (USBD_SUPPORTS_HIGH_SPEED && usbd_bus_speed(uds_ctx) == USBD_SPEED_HS) {
```
Stop pointless reformatting!

nrbrook commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -72,15 +102,47 @@ struct usb_midi_header_descriptor {
+
 /* midi20 5.3.2 Class-Specific MIDI Streaming Data Endpoint Descriptor */
-struct usb_midi_cs_endpoint_descriptor {
+struct usb_midi2_cs_endpoint_descriptor {
```
It was renamed for clarity, to distinguish it from MIDI 1 descriptor.
Following your feedback, restored the existing `struct usb_midi_cs_endpoint_descriptor` name and kept the new MIDI 1.0 endpoint descriptor as a separate type. The rename-only change was removed.

nrbrook commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -153,9 +219,317 @@ struct usbd_midi_data {
+	}
+
+	uint8_t expected_cin;
+	int payload_len = midi1_status_payload_len(byte0, &expected_cin);
```
Moved declarations to the tops of the affected blocks throughout the touched code paths. This cleanup covers the MIDI 1.0 conversion helpers and the send/control paths while keeping the functional MIDI 1.0 support changes intact.

nrbrook commented:
@jfischer-no any chance of re-review/approve/merge? Thanks

carlescufi commented:
@titouanc @jfischer-no can you please take a look?

carlescufi commented:
@nrbrook could you please rebase the PR?

titouanc commented on include/zephyr/audio/midi.h:
```diff
@@ -356,6 +356,53 @@ struct midi_ump {
 /** @} */
 
+/**
+ * @defgroup midi_status_byte MIDI 1.0 Status Bytes
```
I find it confusing that these values are re-defined in here, as they are already defined with other names earlier in this file, see the Doxygen groups above: `midi_ump_cmd` and `midi_ump_sys`

nrbrook replied:
Agreed — the additions duplicated the `midi_ump_cmd` / `midi_ump_sys` Doxygen groups already in this file. Reverted in the new `Switch USB-MIDI public API to UMP-only` commit; the public additions in `<zephyr/audio/midi.h>` are gone, and the few status-byte constants the driver still needs internally now live in `subsys/usb/device_next/class/usbd_midi2.c`.

titouanc commented on include/zephyr/usb/class/usbd_midi2.h:
```diff
@@ -25,6 +25,21 @@ extern "C" {
 #include <zephyr/device.h>
 #include <zephyr/audio/midi.h>
 
+/**
+ * @brief Raw USB-MIDI 1.0 message
+ *
+ * Represents one USB-MIDI 1.0 payload with a virtual cable number and up to
+ * three MIDI bytes.
+ */
+struct usbd_midi1_packet {
+	/** Virtual cable (group) number. */
+	uint8_t cable_number;
+	/** Number of valid bytes stored in @p bytes (1 to 3). */
+	uint8_t len;
+	/** MIDI status byte followed by up to two payload bytes. */
+	uint8_t bytes[3];
+};
```
This is NOT the structure of a USB-MIDI 1.0 packet.

See [Universal Serial Bus Device Class Definition for MIDI Devices](https://www.usb.org/sites/default/files/midi10.pdf) (hereafter designated as midi10.pdf)

In section 4: USB-MIDI Event Packets
<img width="608" height="131" alt="Image" src="https://github.com/user-attachments/assets/e89fd065-6416-4fc5-a29a-dccab642957f" />

In particular, you don't need the length attribute, since it is entirely dependent on the Code Index Number:

<img width="739" height="549" alt="Image" src="https://github.com/user-attachments/assets/f6f2162c-edb3-4849-999e-32253a418b11" />

<img width="747" height="212" alt="Image" src="https://github.com/user-attachments/assets/b79625fd-2a24-4c31-93d2-88f4db74ecb4" />

Finally, keeping it aligned on 4 bytes feels more natural, especially on common 32b MCUs targetted by Zephyr

nrbrook replied:
Agreed — that wasn't the USB-MIDI 1.0 event packet from midi10.pdf §4 and the `len` field was redundant given the CIN. Rather than fix the type, I went with your earlier suggestion and made the public API UMP-only: `struct usbd_midi1_packet`, `usbd_midi_send_midi1()` and `rx_midi1_cb` are all gone in the new `Switch USB-MIDI public API to UMP-only` commit. Applications now send and receive UMPs only; the driver does the MIDI 1.0 ↔ UMP translation internally on the legacy alt setting.

titouanc commented on samples/subsys/usb/midi/README.rst:
```diff
@@ -52,6 +54,23 @@ support; like your favorite Digital Audio Workstation or synthetizer. If you
+automatically converts them to either raw USB-MIDI 1.0 packets or Universal
+MIDI Packets depending on the active alternate. As a result the sample behaves
+identically from the point of view of the host regardless of its MIDI
+capabilities.
```
I find it weird to have the conversion in that order (MIDI1 promoted to MIDI2), especially since the USB-MIDI1.0 interface in this PR only support a single input/single output. Therefore, how is the UMP group supposed to be routed ?

nrbrook replied:
Fair point — the original implementation only wired up a single embedded IN/OUT jack pair on the MIDI 1.0 alt regardless of how many Group Terminal Blocks the device tree declared, so cable 0 was the only thing the host ever saw and the conversion direction was the only one that made sense.

The new `Expose USB-MIDI 1.0 jacks per Group Terminal Block` commit fixes this: each DT child now contributes its own jack set based on `terminal-type` (`bidirectional` → host→device + device→host pair; `output-only` → host→device only; `input-only` → device→host only), and a per-instance compile-time lookup table maps USB-MIDI 1.0 cable numbers to MIDI 2.0 group numbers in each direction. So a device with three GTBs at groups 0/2/5 gets three virtual cables on the legacy alt, and the group numbers survive the round trip even when they're sparse.

I also rewrote this section of the README to describe the new behaviour and the UMP-only application API.

titouanc commented on samples/subsys/usb/midi/src/main.c:
With this update to the `key_press` event handler, it is really unclear to me how the application is supposed to know when the MIDI2 interface is selected.

I understand that there is this conversion from MIDI1 to MIDI2 (UMP) happening automagically, but how is the UMP group routing supposed to work ? How can I send anything else than MIDI1 data (for example, a MIDI2 note-on event with custom attributes) ?

nrbrook replied:
That ambiguity is fixed in the new `Switch USB-MIDI public API to UMP-only` commit. The application API is now uniformly UMP-based — `usbd_midi_send()` takes a UMP regardless of which alt is selected, and the driver translates to USB-MIDI 1.0 event packets internally if the host has the legacy alt active. So `key_press` no longer has to care which alt is current; it just builds a `UMP_MIDI1_CHANNEL_VOICE(...)` and sends it.

For the cases where the application *does* want to observe the alt — e.g. to gate MIDI 2-only payloads such as note-on with attributes when the host is on the legacy alt — I added two pieces:

- `int usbd_midi_get_active_alt(const struct device *dev)` returning a `enum usbd_midi_alt` (`USBD_MIDI_ALT_MIDI1` / `USBD_MIDI_ALT_MIDI2`),
- `usbd_midi_ops.alt_changed_cb`, called whenever the host changes the active alternate setting.

The sample now wires up `alt_changed_cb` purely to log which alt the host has selected, as a usage example. UMP group routing is now driven by the per-GTB cable mapping introduced in the multi-jack commit (replied to separately on the README comment).

titouanc commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -154,17 +234,273 @@ struct usbd_midi_data {
+	default:
+		return 0;
+	}
+}
```
NB: for this you can reuse the same idea as the macro `UMP_NUM_WORDS` https://github.com/zephyrproject-rtos/zephyr/blob/v4.4.0/include/zephyr/audio/midi.h#L81-L97

nrbrook replied:
Good idea, but it turned out not to be needed — the on-the-fly SysEx assembly state in `struct usbd_midi_data` (`sysex_buf`, `sysex_buf_len`, `sysex_transfer_active`) only existed to back the now-removed `rx_midi1_cb` path. With the public API switching to UMP-only, all of that is gone, so there's nothing left to size with `UMP_NUM_WORDS`.

titouanc commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -566,228 +1218,295 @@ void usbd_midi_set_ops(const struct device *dev, const struct usbd_midi_ops *ops
-		.wMaxInputBandwidth = sys_cpu_to_le16(DT_PROP(node, serial_31250bps)),  \
-		.wMaxOutputBandwidth = sys_cpu_to_le16(DT_PROP(node, serial_31250bps)), \
+		.wMaxInputBandwidth = 0x0000,                                           \
+		.wMaxOutputBandwidth = 0x0000,                                          \
```
Why are you removing this ?

nrbrook replied:
That was an unintentional regression on my side — `wMaxInputBandwidth` / `wMaxOutputBandwidth` should still come from `DT_PROP(node, serial_31250bps)`. Restored in the first of the new commits (`Restore Group Terminal Block bandwidth from DT`); both fields are back to `sys_cpu_to_le16(DT_PROP(node, serial_31250bps))`.

titouanc reviewed (commented):
In my humble opinion, there are some issues with this proposal:

- If both USB- MIDI1 and MIDI2 are enabled, then you are limited to single input/output ?
- There is no way for the application to know which alternate is currently enabled (and therefore if it can use UMP or should rely on the legacy packet format)

> The new MIDI1 functions are for raw midi1 data, to simplify implementation for existing applications using midi1 which already have midi1 data in a 3 byte buffer, and avoid unnecessary conversion from/to UMP internally.

- Converting MIDI1 channel voice events (the most common ones) to UMP is trivial and shouldn't really break existing external code
- For SysEx, you already need to do some custom encoding for USB-MIDI1, so why not using UMP anyway ? At least UMP is standard across all transports, and more future-proof.
- You already have partial conversion from/to UMP within the USB-MIDI class implementation. I would rather make it complete, and leverage this to keep the API unified around `usbd_midi_send` and `rx_packet_cb`.

All in all, as I stated in my first review, I would highly prefer the public interface to be based on the UMP format, as it makes everything more consistent

> Currently, this implements one input and one output jack. I did start implementing the configuration of these through device tree, but got lost down a macro rabbit hole, so figured I should get some feedback first.

With all that being said, I would also really like to see backward-compatible USB-MIDI1 landing into Zephyr, such as to support older hosts. I remember fiddling around with those variable-length descriptors and device tree macros, I can possibly give a hand there.

nrbrook commented:
@titouanc thanks for the thorough review — pushed three new commits on top of the previous ones to address the feedback. They're left as separate commits for review readability and will be folded in before merge:

- `usb: device_next: Restore Group Terminal Block bandwidth from DT`
- `usb: device_next: Switch USB-MIDI public API to UMP-only`
- `usb: device_next: Expose USB-MIDI 1.0 jacks per Group Terminal Block`

@carlescufi the branch is now rebased on current `main`.

The headline change is conceding to a UMP-only public API as you suggested. Applications no longer see any MIDI 1.0 type or callback in `<zephyr/usb/class/usbd_midi2.h>`; the MIDI 1.0 alternate setting becomes a pure compatibility layer and the driver translates incoming USB-MIDI 1.0 event packets to UMPs (and outgoing UMPs to event packets) internally. The previously experimental USB-MIDI 1.0 packet type and `rx_midi1_cb` are gone, and the sample no longer needs to know which alternate is active to send notes — it just pushes UMPs and the class converts as needed for the current alt. To keep the alternate observable from the application side (e.g. for limiting outgoing UMP types to MIDI 1.0 channel-voice when the host is on the legacy alt) I added `usbd_midi_get_active_alt()` and an `alt_changed_cb` callback in `usbd_midi_ops`; the sample uses the callback purely to log which alt the host selected.

I also took the opportunity to expose USB-MIDI 1.0 jacks per Group Terminal Block, which addresses the "single virtual cable regardless of how many MIDI 2.0 groups" issue called out in the README/main.c comments. Each DT child now contributes jacks based on its `terminal-type`:

- `bidirectional` → embedded IN + external OUT (host→device) and external IN + embedded OUT (device→host),
- `output-only` → only the host→device pair,
- `input-only` → only the device→host pair.

A small per-instance cable-to-group lookup table is built at compile time in DT child order for each direction, and `midi1_event_to_ump` / `midi1_ump_to_event` take the resolved group / cable as an explicit argument, so MIDI 2.0 group numbers (which can be non-contiguous) survive the round trip. The existing `BUILD_ASSERT(DT_REG_SIZE(node) == 1, ...)` is kept for now; supporting GTBs of size > 1 needs N jack sets per block and is straightforward to add as a follow-up if there's demand.

I tested the result on `nrf52840dk/nrf52840`. With the default single-GTB sample the device enumerates with both alt settings, CoreMIDI (on Mac) presents both as ports, and a 9-message channel-voice loopback round-trips cleanly on each of them. Repeated the same test with a 4-GTB overlay (bidirectional + output-only + input-only + bidirectional); the resulting MIDI 1.0 alternate descriptor contains the expected 12 jacks (six embedded/external pairs in each direction, with `baSourceID` wired between matched pairs) and the loopback passes 9/9 on both alt-setting ports.

nrbrook commented:
I would like it removed, but I'm losing faith this PR will ever be merged, sadly

josuah reviewed (commented):
In order to prepare the upcoming reviews, would it be possible to reorganize  the commits so that they only introduce new content?

For instance:

> Clean up the review-fix commit

Zephyr tries to avoid "clean-up" commits and encourages introducing commits that are immediately having the final version. (ctrl+f "fixup" on <https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-requirements>)

Thank you for your patience!

nrbrook commented:
Yes I can do, I mentioned I would do this before merge, it was left this way for review so the recent changes could be viewed in isolation.

titouanc commented:
Hello @nrbrook and sorry for the delay again.

I just run a quick test on your PR and at least the USB-MIDI1 descriptors look correct :+1: . Since my host PC has built-in support for USB-MIDI2, it automatically enables the MIDI2 alternate, so I have to look at how to downgrade it to MIDI1 to test further. I will try to look into this, as well as the code itself in the next days.

By the way, shouldn't it be a kconfig option to independently enable or disable USB-MIDI 1/2 ?

> it was left this way for review so the recent changes could be viewed in isolation.

I am not a maintainer, so I can not comment on the commit workflow, however as @josuah pointed out you should aim at a clean commit history. I understand you try to highlight the changes since the last review, but this is not really important; what matters the most is the final state of the changes and their accompanying commit message. If you get any review now, they will anyway be dismissed as soon as you fixup your commits, so you will need another round of reviews

nrbrook commented:
@titouanc 

> By the way, shouldn't it be a kconfig option to independently enable or disable USB-MIDI 1/2 ?

It was, I removed this because of @jfischer-no saying two alts are always required. https://github.com/zephyrproject-rtos/zephyr/pull/99522#discussion_r2955159802

This is another reason I have not yet merged follow up commits - I wasn't 100% that I should remove the ability to selectively enable midi1/2. But as mentioned, I'm happy to roll the follow up commits in when ready.

titouanc commented:
Hello again @nrbrook

I tried the sample using MIDI1 only. For the record, to disable USB-MIDI2 on Linux, I just reloaded the usb midi driver with an option:
```bash
sudo modprobe -r snd-usb-audio
sudo modprobe snd-usb-audio midi2_enable=0
```

And I was able to send/receive plain MIDI1, and see the corresponding logs from the Zephyr app. I still have to look at your latest code changes, but that's already nice :+1:

nrbrook commented:
Thanks @titouanc 

Do you have a view on if compile time/run time midi 1/2 toggles should be available, or are you happy with the current no-config approach?

DaAwesomeP commented:
Just my 2 cents: I 100% support the idea of only having the MIDI 2.0/UMP API and all MIDI 1.0 packets natively converted. This means there are fewer API methods to maintain, the UMP implementation will be continuously validated, and enabling MIDI 2.0 features will be seamless for applications.

In terms of limiting to only a MIDI 1.0 or MIDI 2.0 exposed interface, I think it would be good to start off with how you have implemented it now: both enabled with automatic fallback (as the specification says). While I can see why a product manufacturer might only want to enable MIDI 1.0 or an experimenter might only want MIDI 2.0, it shouldn't technically matter. I think of it the same way that you wouldn't lock a USB 2,0 port to 1.1 to use a 1.1 device. You just need to detect which mode it is in (which you have implemented).

If someone discovers a quirk with an another device or really needs it, then they will open an issue or PR. I think it is worth waiting and seeing--if nobody needs it, then we are saving some configuration complexity. Plus, it will help to proliferate MIDI 2.0 adoption.

nrbrook commented:
Ok thanks. 

> While I can see why a product manufacturer might only want to enable MIDI 1.0 or an experimenter might only want MIDI 2.0, it shouldn't technically matter.

> If someone discovers a quirk with an another device or really needs it, then they will open an issue or PR.

That was the reason I implemented it - my client had midi 1 support with the old usb class api, didn't ask for midi 2, and windows at the time (last year) was not behaving correctly with midi 1+2 (since fixed, but has everyone updated? Are there other OSes with issues?)

So I had implemented it, which exists in the commits as they are right now, and then removed it in a follow up commit. So if there is going to be some demand for it (including from my client) it feels to me like it should be kept... but with default behaviour as it is now.

diegoherranz commented:
My 2 cents as an user who plans to use USB MIDI in Zephyr:

- I'm OK with having USB MIDI 2.0 with USB MIDI 1.0 as fallback with no configuration options.
- I'm OK having only the MIDI 2.0/UMP API for everything. In fact it would be good if there was a backend for old-style MIDI DIN (31250 bps async/UART) and that would still use the same API so that you could use USB or DIN with the same API (obviously for the overlapping capabilities only).

titouanc commented on include/zephyr/usb/class/usbd_midi2.h:
```diff
@@ -42,15 +69,39 @@ struct usbd_midi_ops {
+	 * this to gate behaviour that has no MIDI 1.0 representation (such as
+	 * MIDI 2.0 channel voice messages).
+	 */
+	void (*alt_changed_cb)(const struct device *dev, const enum usbd_midi_alt alt);
```
In my initial USB-MIDI2.0 PR, I had a single callback for readiness+altsetting selection, but since MIDI1.0 was not implemented, we ditched it in favor of the `ready_cb` (see remains of the discussion here: https://github.com/zephyrproject-rtos/zephyr/pull/81197#discussion_r1931283418)

I believe it is better to have a single `status_changed_cb(dev, DISABLED / MIDI1 / MIDI2)`, that would replace both `ready_cb` and `alt_changed_cb`. This avoids firing 2 distinct callbacks when the interface is brought up, and reflects the actual status of the USB-MIDI interface. This also reduces the complexity on the user side (you only have to handle 1 callback to know in which state the interface is).

Since the USB-MIDI API is still in v0.1.0 we can do this without going through the whole "breaking changes" formal process

nrbrook replied:
Done — replaced both `ready_cb` and `alt_changed_cb` with a single `status_changed_cb(dev, status)`, where `status` is one of `USBD_MIDI_STATUS_DISABLED` / `USBD_MIDI_STATUS_MIDI1` / `USBD_MIDI_STATUS_MIDI2`. A single callback now reports both readiness (status != DISABLED) and the active alternate, and it's deduplicated so bringing the interface up fires it once rather than emitting separate ready + alt-changed events. `usbd_midi_get_active_alt()` became `usbd_midi_get_status()` returning the same enum. Folded into the API commit.

titouanc commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -23,7 +26,7 @@ LOG_MODULE_REGISTER(usbd_midi2, CONFIG_USBD_MIDI2_LOG_LEVEL);
 		    sizeof(struct udc_buf_info), NULL);
 
-#define MIDI_QUEUE_SIZE 64
+#define MIDI_QUEUE_SIZE CONFIG_USBD_MIDI2_TX_QUEUE_SIZE
```
Nitpick: replace uses of `MIDI_QUEUE_SIZE` with `CONFIG_USBD_MIDI2_TX_QUEUE_SIZE` instead of aliasing here (AFAIK there's only the buffer declaration that uses that value)

nrbrook replied:
Done — dropped the `MIDI_QUEUE_SIZE` alias; the buffer declaration and `ring_buf_init()` now use `CONFIG_USBD_MIDI2_TX_QUEUE_SIZE` directly. Folded into the "Add Kconfig option for MIDI queue size" commit.

titouanc commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -156,15 +276,249 @@ struct usbd_midi_data {
+	ring_buf_reset(&data->tx_queue);
+}
+
+static int midi1_status_payload_len(const uint8_t status, uint8_t *const cin)
```
Somehow refer to the reference document here, ie.
"midi10: Table 4-1: Code Index Number Classifications"

nrbrook replied:
Done — added `see midi10 Table 4-1: Code Index Number Classifications` to the doc comment on `midi1_status_payload_len()`.

titouanc commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -156,15 +276,249 @@ struct usbd_midi_data {
+
+		ump->data[0] = (UMP_MT_DATA_64 << 28) | ((group & 0xF) << 24) |
+			       ((status & 0xF) << 20) | ((len & 0xF) << 16) | (byte0 << 8) | byte1;
+		ump->data[1] = byte2 << 24;
+
+		return 0;
+	}
```
I haven't tested this properly, but it looks like this is incorrect. Indeed, from the spec, the leading F0 (start of sysex) byte shouldn't be included in the payload, and the same for the trailing F7 (end of sysex)

<img width="741" height="590" alt="Image" src="https://github.com/user-attachments/assets/b778f1d7-341c-4a28-9a01-393a90942b70" />

For example, given the following MIDI1 sysex:

`F0 01 02 03 04 05 F7`

This gives the following USB-MIDI1 packets:

```
1. CIN=0x4 (sysex-start)  F0 01 02
2. CIN=0x4 (sysex-start)  03 04 05
3. CIN=0x5 (sysex-end1b)  F7
```

The current implementation will produce:

```
1. MT=3 status=0x1 (start)    len=3  F0 01 02
2. MT=3 status=0x2 (continue) len=3  03 04 05
3. MT=3 status=0x3 (end)      len=1  F7
```

In an ideal world, I would expect it to produce a single UMP
```
1. MT=3 status=0x0 (complete) len=5  01 02 03 04 05
```

Although the spec says:
> Note: Each System Exclusive UMP may contain fewer than 6 bytes of data. A Start or Continue with fewer than 6 bytes does not signify a message end

So if we don't want to accumulate multiple MIDI1 message in the UMP translation layer, based on the USB-MIDI1 packets we get, we could also emit:

```
1. MT=3 status=0x1 (start)    len=2  01 02
2. MT=3 status=0x2 (continue) len=3  03 04 05
3. MT=3 status=0x3 (end)      len=0
```

titouanc replied:
As a side note, this sysex translation is complex enough that we could extract it to its own function 😉

nrbrook replied:
Good catch — fixed. The F0/F7 framing bytes are now stripped from the UMP payload on the way in and re-inserted on the way out. For your example `F0 01 02 03 04 05 F7` the driver now produces `start len=2 [01 02] / continue len=3 [03 04 05] / end len=0` (your option B), and that round-trips back to the original USB-MIDI 1.0 packets. The TX worst-case reservation was bumped to 12 bytes (6 payload + F0/F7 -> 3 packets) to account for the added framing.

nrbrook replied:
Done — the SysEx translation now lives in dedicated helpers: `midi1_sysex_event_to_ump()` for the incoming (USB-MIDI 1.0 -> UMP) direction and `midi1_ump_to_sysex_events()` for the outgoing one.

titouanc commented:
> Do you have a view on if compile time/run time midi 1/2 toggles should be available, or are you happy with the current no-config approach?

Having a KConfig option to optionally enable the USB-MIDI2.0 interface could be interesting, as it would make testing of USB-MIDI1.0 much easier. Note that USB-MIDI1.0 on altsetting 0 is always required to enumerate properly (although it can be "empty" like in the implementation from the original USB-MIDI2 PR #81197).

In any case, I don't consider this as a blocker, and we could easily add that in a subsequent PR :slightly_smiling_face: 

> Just my 2 cents: I 100% support the idea of only having the MIDI 2.0/UMP API and all MIDI 1.0 packets natively converted. This means there are fewer API methods to maintain, the UMP implementation will be continuously validated, and enabling MIDI 2.0 features will be seamless for applications.

Indeed, this was my primary concern. Thank you for applying that @nrbrook :+1:

nrbrook committed:
usb: device_next: Add MIDI 1.0 support to the USB next MIDI 2.0 driver

Add backward-compatible USB-MIDI 1.0 support to the existing USB MIDI 2.0
device_next class driver, so the device still enumerates and exchanges
MIDI with hosts that only speak the legacy USB-MIDI 1.0 protocol.

The public application API stays Universal MIDI Packet (UMP) only:
applications send and receive UMPs through usbd_midi_send and
usbd_midi_ops.rx_packet_cb regardless of which alternate setting the host
has selected, and the driver transparently translates between USB-MIDI
1.0 event packets and UMPs internally.

USB-MIDI 1.0 jack descriptors are generated per device-tree child /
Group Terminal Block rather than a single hard-coded jack pair:

  - "bidirectional" GTBs contribute a host-to-device pair (embedded IN
    jack + external OUT jack) and a device-to-host pair (embedded OUT
    jack + external IN jack).
  - "output-only" GTBs contribute only the host-to-device pair.
  - "input-only" GTBs contribute only the device-to-host pair.

A small per-instance cable-to-group lookup table is kept for each
direction so that MIDI 2.0 group numbers (which can be non-contiguous)
are preserved when translating between USB-MIDI 1.0 cable numbers and
UMP groups.

Applications that need to gate features with no MIDI 1.0 representation
(for example UMP_MT_MIDI2_CHANNEL_VOICE) can observe the interface
status (disabled, or the active alternate) via enum usbd_midi_status,
usbd_midi_ops.status_changed_cb and usbd_midi_get_status.

Signed-off-by: Nick Brook <nrbrook@gmail.com>

nrbrook committed:
samples: subsys: usb: midi: Update sample for USB MIDI 1.0 support

Update the USB MIDI sample application and its README to use the unified
UMP application API and to track the interface status via
usbd_midi_ops.status_changed_cb, demonstrating operation over both the
USB-MIDI 1.0 and 2.0 alternate settings.

Signed-off-by: Nick Brook <nrbrook@gmail.com>

nrbrook committed:
usb: device_next: Added MIDI class documentation

Add documentation for the USB device_next MIDI class, covering the
UMP-based application API and the USB-MIDI 1.0 / 2.0 alternate settings.

Signed-off-by: Nick Brook <nrbrook@gmail.com>

nrbrook committed:
usb: device_next: Add Kconfig option for MIDI queue size

Add a Kconfig option for the MIDI transmit queue size,
CONFIG_USBD_MIDI2_TX_QUEUE_SIZE, instead of a hard-coded value.

Signed-off-by: Nick Brook <nrbrook@gmail.com>

nrbrook commented:
Pushed an update addressing the review feedback:

- **Unified the callbacks** — `ready_cb` + `alt_changed_cb` are now a single `status_changed_cb(dev, status)` with `enum usbd_midi_status` (`DISABLED`/`MIDI1`/`MIDI2`); `usbd_midi_get_active_alt()` → `usbd_midi_get_status()`.
- **SysEx F0/F7 framing** — framing bytes are now stripped from the UMP payload on RX and re-inserted on TX, and the translation is factored into `midi1_sysex_event_to_ump()` / `midi1_ump_to_sysex_events()`.
- Dropped the `MIDI_QUEUE_SIZE` alias (use `CONFIG_USBD_MIDI2_TX_QUEUE_SIZE` directly) and added the `midi10 Table 4-1` spec reference on `midi1_status_payload_len()`.

I also cleaned up the history as requested: the branch is now 4 logical commits (core MIDI 1.0 support, docs, sample, Kconfig queue size). The runtime MIDI 1.0/2.0 alt-setting toggle has been dropped from this PR and will come as a follow-up. Each review fix was folded into the relevant commit, so this is a force-push.

DaAwesomeP commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -141,30 +250,358 @@ struct usbd_midi_descriptors {
+		if (cin != NULL) {
+			*cin = MIDI_CIN_SYSEX_END_1BYTE;
+		}
+		return 1;
+	case MIDI_STATUS_SYSEX_END:
+		if (cin != NULL) {
+			*cin = MIDI_CIN_SYSEX_END_1BYTE;
+		}
+		return 1;
```
```suggestion
	case MIDI_STATUS_TUNE_REQUEST:
	case MIDI_STATUS_SYSEX_END:
		if (cin != NULL) {
			*cin = MIDI_CIN_SYSEX_END_1BYTE;
		}
		return 1;
```

nrbrook replied:
Thanks! Applied — merged the two identical `MIDI_STATUS_TUNE_REQUEST`/`MIDI_STATUS_SYSEX_END` cases into a single fall-through.

DaAwesomeP commented on samples/subsys/usb/midi/src/main.c:
```diff
@@ -65,17 +84,33 @@ static void on_midi_packet(const struct device *dev, const struct midi_ump ump)
+	case USBD_MIDI_STATUS_MIDI1:
+		LOG_INF("Host selected USB-MIDI 1.0 alternate setting");
+		break;
+	case USBD_MIDI_STATUS_MIDI2:
+		LOG_INF("Host selected USB-MIDI 2.0 alternate setting");
+		break;
```
```suggestion
	case USBD_MIDI_STATUS_MIDI2:
		LOG_INF("Host selected USB-MIDI 2.0 alternate setting");
		break;
	default:
	    CODE_UNREACHABLE;
```

nrbrook replied:
Thanks! Applied — added `default: CODE_UNREACHABLE;` to the status switch.

DaAwesomeP reviewed (changes requested):
We can easily clear the more important Sonarqube issues.

nrbrook commented:
Addressed the open SonarQube issues in the latest push:

- **`usbd_midi2.c`** – merged the duplicate `MIDI_STATUS_TUNE_REQUEST`/`MIDI_STATUS_SYSEX_END` `case` blocks into a single fall-through.
- **`usbd_midi2.c`** – declared the loop counter `i` inside the `for` loops in `usbd_midi2_recv()` and `usbd_midi_send()`.
- **`samples/.../main.c`** – added a `default: CODE_UNREACHABLE;` to the status switch.

Rebased on latest `main`; builds clean and compliance passes (the only remaining ClangFormat notes are pre-existing macro-alignment warnings from a newer local clang-format version, advisory only).

DaAwesomeP commented:
Seems like this is ready to go?

nrbrook commented:
@DaAwesomeP  I think so! Who else needs to approve it?

DaAwesomeP commented:
> @DaAwesomeP I think so! Who else needs to approve it?

You might try pinging the assignees in the `#pr-help` channel on Discord. You can also click the re-request review button next to their names at the top of the PR.

titouanc commented on include/zephyr/usb/class/usbd_midi2.h:
```diff
@@ -33,24 +58,49 @@ struct usbd_midi_ops {
+	 * MIDI 2.0 channel voice messages).
 	 */
-	void (*ready_cb)(const struct device *dev, const bool ready);
+	void (*status_changed_cb)(const struct device *dev, enum usbd_midi_status status);
```
NB: this is a breaking API change, although the USB-MIDI API is still in v0.1.0

At the very least, I would bump the API to v0.2.0, and possibly add a note in the migration guide for Zephyr v4.5 (cc @kartben)

nrbrook replied:
Done — bumped the USB MIDI class API to v0.2.0 and added a note in the Zephyr v4.5 migration guide covering the `ready_cb` → `status_changed_cb` / `usbd_midi_status` change.

titouanc commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -141,21 +250,344 @@ struct usbd_midi_descriptors {
+	bytes[3] = (payload_len == 2) ? 0x00 : UMP_MIDI1_P2(*ump);
+
+	*event_le = (uint32_t)bytes[0] | ((uint32_t)bytes[1] << 8) | ((uint32_t)bytes[2] << 16) |
+		    ((uint32_t)bytes[3] << 24);
```
```suggestion
	*event_le = sys_get_le32(bytes);
```

nrbrook replied:
Done — packing the USB-MIDI 1.0 event now uses `sys_get_le32(bytes)`.

titouanc commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -455,34 +899,193 @@ static int usbd_midi_preinit(const struct device *dev)
+				     uint32_t events[3])
+{
+	uint8_t status = (ump->data[0] >> 20) & 0xF;
+	uint8_t len = (ump->data[0] >> 16) & 0xF;
```
IMHO, this deserves proper defines in `include/zephyr/audio/midi.h`, for example:

```C
#define UMP_DATA64_STATUS(ump) (((ump).data[0] >> 20) & 0xF)
#define UMP_DATA64_LEN(ump)    (((ump).data[0] >> 16) & 0xF)
```

nrbrook replied:
Done — added `UMP_DATA64_STATUS()` and `UMP_DATA64_LEN()` in `include/zephyr/audio/midi.h` and switched the driver over to them.

titouanc commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -455,34 +899,193 @@ static int usbd_midi_preinit(const struct device *dev)
+		return -EINVAL;
+	}
+
+	payload[0] = (ump->data[0] >> 8) & 0xFF;
+	payload[1] = ump->data[0] & 0xFF;
+	payload[2] = (ump->data[1] >> 24) & 0xFF;
+	payload[3] = (ump->data[1] >> 16) & 0xFF;
+	payload[4] = (ump->data[1] >> 8) & 0xFF;
+	payload[5] = ump->data[1] & 0xFF;
```
As we are emitting MIDI1, we should ensure that data is only 7b wide.

There are 2 ways:
1. mask to 7 here: `ump->data[1] & 0x7F` (and silently erase the most significant bit if it had been set)
2. return an error (-EINVAL) if any of the data bytes has the most significant bit set

I am personally in favor of option 2.

nrbrook replied:
Went with option 2: `midi1_ump_to_sysex_events()` now returns `-EINVAL` if any of the valid SysEx payload bytes has the high bit set.

titouanc commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -455,34 +899,193 @@ static int usbd_midi_preinit(const struct device *dev)
+			uint32_t events[3];
+			int n_events;
+
+			n_events = midi1_ump_to_sysex_events(&ump, cable_number, events);
+			if (n_events < 0) {
+				return n_events;
+			}
```
Nitpick: if we perform that conversion earlier, we wouldn't need to reserve the worst case 12B on line 1043 above.

But that's already quite an edge case (sending a short Sysex in an almost full Tx buffer), so not a problem if you do not address it in this PR 😉

nrbrook replied:
Addressed as well — the UMP is converted to USB-MIDI 1.0 packets first, then we reserve the actual encoded size instead of a worst-case 12 bytes.

toonst commented:
I ran into an issue trying this out:

### `MIDI1_MS_TOTAL_LEN()` overstates `wTotalLength`, which crashes CoreMIDI on macOS

The MidiStreaming interface header's `wTotalLength` counts the two standard endpoint descriptors and the two class-specific endpoint descriptors:

```c
#define MIDI1_MS_TOTAL_LEN(n)                                                  \
	(sizeof(struct usb_midi_header_descriptor) +                           \
	 N_OUTPUTS(n) * sizeof(struct usb_midi_in_jack_descriptor) +           \
	 N_OUTPUTS(n) * sizeof(struct usb_midi_out_jack_descriptor) +          \
	 N_INPUTS(n) * sizeof(struct usb_midi_in_jack_descriptor) +            \
	 N_INPUTS(n) * sizeof(struct usb_midi_out_jack_descriptor) +           \
	 2 * sizeof(struct usb_ep_descriptor) +   /* <-- not part of the field */ \
	 (4 + N_OUTPUTS(n)) +                     /* <-- not part of the field */ \
	 (4 + N_INPUTS(n)))                       /* <-- not part of the field */
```

midi10 6.1.2.1 defines the field as *"the combined length of this descriptor header and all Jack and Element descriptors"* — endpoint descriptors are not included. For comparison, TinyUSB's `TUD_MIDI_DESCRIPTOR` uses `7 + numcables * 30`.

#### Effect

With one bidirectional Group Terminal Block, this declares **61** where the header plus jacks total **37**. Decoded from the built image:

```
MS_HEADER      len=7
MIDI_IN_JACK   len=6      declared wTotalLength : 61
MIDI_OUT_JACK  len=9      header + jacks        : 37
MIDI_IN_JACK   len=6
MIDI_OUT_JACK  len=9
```

The extra 24 bytes are exactly the two endpoint descriptors and their two class-specific endpoint descriptors, which physically follow in the configuration descriptor. A host walking the class-specific block by `wTotalLength` therefore runs into standard `ENDPOINT` (0x05) descriptors in a region where only Jack and Element descriptors are valid.

On macOS this segfaults `AppleMIDIUSBDriver`, which takes `MIDIServer` down with it:

```
exception:   EXC_BAD_ACCESS (SIGSEGV), KERN_INVALID_ADDRESS at 0x0000000000000110
   AppleMIDIUSBDriver           ?
   AppleMIDIUSBDriver           ?
   CoreMIDI                     CADeprecated::TimerTaskMgr::Run()
```

Audio MIDI Setup then reports *"The MIDI server cannot be opened"* and **no** MIDI device on the system is usable until MIDIServer is restarted — so a single malformed device takes out all MIDI I/O, not just its own.

#### Fix

```diff
 #define MIDI1_MS_TOTAL_LEN(n)                                                  \
 	(sizeof(struct usb_midi_header_descriptor) +                           \
 	 N_OUTPUTS(n) * sizeof(struct usb_midi_in_jack_descriptor) +           \
 	 N_OUTPUTS(n) * sizeof(struct usb_midi_out_jack_descriptor) +          \
 	 N_INPUTS(n) * sizeof(struct usb_midi_in_jack_descriptor) +            \
-	 N_INPUTS(n) * sizeof(struct usb_midi_out_jack_descriptor) +           \
-	 2 * sizeof(struct usb_ep_descriptor) +                                \
-	 (4 + N_OUTPUTS(n)) +                                                  \
-	 (4 + N_INPUTS(n)))
+	 N_INPUTS(n) * sizeof(struct usb_midi_out_jack_descriptor))
```

With this applied the device enumerates cleanly, MIDIServer is stable, and MIDI 1.0 in/out works.

#### Notes

- **Pre-existing, same root cause:** `if1_0_ms_header` on `main` computes `sizeof(header) + 2 * (sizeof(usb_ep_descriptor) + 4)` = 29 for a block whose header is its only class-specific descriptor (7). It is harmless today only because that alternate carries no jacks and is never functionally parsed, but it is the same misreading and is worth correcting in the same patch.
- **Exposure:** I hit this with `CONFIG_USBD_MIDI2_ALTSETTING_MIDI2=n` (from your `midi2-add-midi1-toggles` branch), where macOS has no MIDI 2.0 alternate to select and parses the MIDI 1.0 block. I have not tested it, but I would expect any MIDI-1-only host to parse the same block even with both alternates present — which is precisely the audience this PR targets.
- An overshooting length field is benign in parsers that bound-check against the actual descriptor buffer, which may be why this survived testing on other hosts.

#### Environment

- Zephyr: `midi2-add-midi1-toggles` @ `986799e2eb6`
- Board: custom RP2040, one `zephyr,midi2-device` child, `reg = <0 1>`, `terminal-type` bidirectional
- Host: macOS

nrbrook commented:
@toonst Thanks for the excellent write-up and the macOS CoreMIDI repro.

Fixed `MIDI1_MS_TOTAL_LEN()` so `wTotalLength` is only the MS header plus Jack descriptors (midi10 6.1.2.1), not the endpoint descriptors. For a single bidirectional GTB that is 37 rather than 61.

The dummy alt-0 header on `main` is replaced in this PR by the real MIDI 1.0 jack block, so correcting this macro covers that case as well.

tmon-nordic commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -52,6 +53,71 @@ UDC_BUF_POOL_DEFINE(usbd_midi_buf_pool, DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) *
+#define MIDI_CIN_CHANNEL_PRESSURE	0xD
+#define MIDI_CIN_PITCH_BEND_CHANGE	0xE
+#define MIDI_CIN_SINGLE_BYTE		0xF
+
+/* MIDI 1.0 status bytes used internally to classify USB-MIDI 1.0 events.
+ * see midi10 Appendix 1.1 Table of MIDI 1.0 Messages
+ */
```
Can you name precisely what document this is from? I cannot find it in "Universal Serial Bus Device Class Definition for MIDI Devices Release 1.0 Nov 1, 1999".

Is this referring to "Standard MIDI-File Format Spec. 1.1 [Appendix 1.1](https://midimusic.github.io/tech/midispec.html#BMA1_1) - Table of Major MIDI Messages" or something else?

nrbrook replied:
You're right — that citation was wrong. These are MIDI 1.0 status bytes, not USB-MIDI class codes, and they are not in midi10. Updated the comment to `ump112` 7.3 / 7.6, which is the in-tree reference for the same values (`include/zephyr/audio/midi.h`).

tmon-nordic commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -52,6 +53,71 @@ UDC_BUF_POOL_DEFINE(usbd_midi_buf_pool, DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) *
+#define MIDI_STATUS_SYSEX_END        0xF7
+#define MIDI_STATUS_TIMING_CLOCK     0xF8
+
+/* midi10 6.1 MIDI Streaming Class-Specific Descriptor constants */
```
Is it more Appendix A. Audio Device Class Codes: MIDIStreaming?

nrbrook replied:
Yes — these are Appendix A codes, not section 6.1. Split to match the existing `midi20 A.*` comments: `midi10 A.1` for the jack subtypes and `midi10 A.3` for embedded/external jack types.

tmon-nordic commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -52,6 +53,71 @@ UDC_BUF_POOL_DEFINE(usbd_midi_buf_pool, DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) *
+
+#define MIDI1_EVENT_BYTES 4
+
+/* ump128 7.7.2 System Exclusive 7-bit Data Messages */
```
Which document is this?

nrbrook replied:
That was a typo. There is no `ump128` document here — the UMP spec used in-tree is `ump112` (M2-104-UM 1.1.2). Changed this (and the other two `ump128` comments in this file) to `ump112` 7.7.

tmon-nordic commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -72,6 +138,15 @@ struct usb_midi_header_descriptor {
 	uint16_t wTotalLength;
 } __packed;
 
+/* midi10 6.1.2 Class-Specific Bulk Endpoint Descriptor */
```
```suggestion
/* midi10 6.2.2 Class-Specific MS Bulk Data Endpoint Descriptor */
```

nrbrook replied:
Applied — this is midi10 6.2.2 Class-Specific MS Bulk Data Endpoint Descriptor.

tmon-nordic commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -81,6 +156,29 @@ struct usb_midi_cs_endpoint_descriptor {
 	uint8_t baAssoGrpTrmBlkID[16];
 } __packed;
 
+/* midi10 6.1.1 MIDI IN Jack Descriptor */
```
```suggestion
/* midi10 6.1.2.2 MIDI IN Jack Descriptor */
```

nrbrook replied:
Applied — midi10 6.1.2.2 MIDI IN Jack Descriptor.

tmon-nordic commented on subsys/usb/device_next/class/usbd_midi2.c:
```diff
@@ -81,6 +156,29 @@ struct usb_midi_cs_endpoint_descriptor {
+	uint8_t iJack;
+} __packed;
+
+/* midi10 6.1.2 MIDI OUT Jack Descriptor (single input pin) */
```
```suggestion
/* midi10 6.1.2.3 MIDI OUT Jack Descriptor (single input pin) */
```

nrbrook replied:
Applied — midi10 6.1.2.3 MIDI OUT Jack Descriptor.

tmon-nordic reviewed (commented):
Are there different competing USB MIDI specification floating around? Things that are referenced, or in case of wTotalLength it sound like copied outright from the spec do not appear in my specification.
